bluetooth: host: Fix uninitialized own_addr_type for legacy scan+adv

olivier-le-sage opened the pull request:
In 25c993e5b74562137c2951db01c8a8d200e98d3d a new case was introduced where own_addr_type is not set by bt_id_set_scan_own_addr properly.

This led to issues for users where increasing their zephyr version led to failures to start scanning after advertising in the case where CONFIG_BT_SCAN_WITH_IDENTITY=n and legacy advertising commands are used.

olivier-le-sage committed:
bluetooth: host: rename bool scan_enabled -> scan_disabled

It represents whether scanning was disabled as part of this flow.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>

Thalley commented on subsys/bluetooth/host/id.c:
```diff
@@ -1840,20 +1840,22 @@ int bt_id_set_scan_own_addr(bool active_scan, uint8_t *own_addr_type)
+				}
+			} else {
+				err = bt_id_set_private_addr(BT_ID_DEFAULT);
+				if (err) {
```
```suggestion
				if (err != 0) {
```

Thalley commented on subsys/bluetooth/host/id.c:
```diff
@@ -1969,21 +1971,31 @@ int bt_id_set_adv_own_addr(struct bt_le_ext_adv *adv, uint32_t options,
 			    !IS_ENABLED(CONFIG_BT_SCAN_WITH_IDENTITY) &&
 			    atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING)) {
-				scan_enabled = true;
+				scan_disabled = true;
```
yikes, this seems like a very old bug :o

Thalley replied:
We should probably conditionally set `scan_disabled` based on the return value of `bt_le_scan_set_enable`

olivier-le-sage replied:
have now added this as a separate commit

Thalley commented on subsys/bluetooth/host/id.c:
```diff
@@ -1980,8 +1980,18 @@ int bt_id_set_adv_own_addr(struct bt_le_ext_adv *adv, uint32_t options,
 			}
-#endif /* defined(CONFIG_BT_OBSERVER) */
+
+			/* If we are scanning with the identity address, it does
+			 * not make sense to set an NRPA.
+			 */
+			if (!(IS_ENABLED(CONFIG_BT_SCAN_WITH_IDENTITY) &&
+			    atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING))) {
+				err = bt_id_set_adv_private_addr(adv);
+			}
```
Is this the right place to have this?

Based on the naming of this function, it relates to advertisements, so why are handling a case for `CONFIG_BT_SCAN_WITH_IDENTITY` here?

Thalley replied:
I also why `CONFIG_BT_SCAN_WITH_IDENTITY` even exists as a Kconfig instead of being a `bt_le_scan_opt` value, since it's just changing some runtime behavior anyways :o

olivier-le-sage commented:
> @olivier-le-sage thanks for the PR. Does this fix this #85202 ?

Ah, I didn't see that bug ticket. Yes, I think this fixes that, unless there is a case where `bt_dev.id_addr[BT_ID_DEFAULT].type` can be neither `BT_ADDR_LE_RANDOM` nor `BT_ADDR_LE_PUBLIC`.

cvinayak commented on subsys/bluetooth/host/id.c:
```diff
@@ -1969,21 +1971,31 @@ int bt_id_set_adv_own_addr(struct bt_le_ext_adv *adv, uint32_t options,
 			 */
 #if defined(CONFIG_BT_OBSERVER)
-			bool scan_enabled = false;
+			bool scan_disabled = false;
```
```suggestion
			bool scan_enabled = false;
```

This is the shadow variable to remember that scanning is active. And it needs to be enabled after the address has been updated (after stopping scanning), as Spec does not permit address update if advertising or scanning is active.

cvinayak replied:
You are testing `atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING)`  and setting `scan_disabled = true`. How is this readable ?

cvinayak replied:
`scan_was_enabled` may be better than `scan_enabled`... 

but again this naming change (commit) is outside scope of this PR, should be a separate PR to make the PR easier to review under its actual scope.

olivier-le-sage replied:
> You are testing atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING) and setting scan_disabled = true. How is this readable ?

so if you look on the next line you'll see the scanner is being disabled there. That's all I was trying to express.

olivier-le-sage commented on subsys/bluetooth/host/id.c:
```diff
@@ -1840,20 +1840,22 @@ int bt_id_set_scan_own_addr(bool active_scan, uint8_t *own_addr_type)
-			if (err) {
-				if (active_scan || !is_adv_using_rand_addr()) {
+		if (!IS_ENABLED(CONFIG_BT_SCAN_WITH_IDENTITY)) {
+			if (is_adv_using_rand_addr()) {
+				if (bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_RANDOM) {
+					*own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
+				} else {
+					*own_addr_type = BT_HCI_OWN_ADDR_PUBLIC;
+				}
+			} else {
```
I don't intend to modify the controller. What do you mean exactly?

olivier-le-sage replied:
I'm not aware of an issue like what you're describing. But I also don't understand what it is you're trying to communicate

cvinayak commented on subsys/bluetooth/host/id.c:
```diff
@@ -1840,20 +1840,22 @@ int bt_id_set_scan_own_addr(bool active_scan, uint8_t *own_addr_type)
+				if (bt_dev.id_addr[BT_ID_DEFAULT].type == BT_ADDR_LE_RANDOM) {
+					*own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
+				} else {
+					*own_addr_type = BT_HCI_OWN_ADDR_PUBLIC;
```
Hmmm... with `is_adv_using_rand_addr()` == `true`  .... we have `*own_addr_type = BT_HCI_OWN_ADDR_PUBLIC;` ?

olivier-le-sage replied:
is it not possible for the scanner to use the public device address while the advertiser uses the random device address?

`is_adv_using_rand_addr()` also only checks `BT_ADV_ENABLED`

olivier-le-sage commented on subsys/bluetooth/host/id.c:
```diff
@@ -1969,21 +1971,31 @@ int bt_id_set_adv_own_addr(struct bt_le_ext_adv *adv, uint32_t options,
+			/* If we are scanning with the identity address, it does
+			 * not make sense to set an NRPA.
+			 */
+			if (!(IS_ENABLED(CONFIG_BT_SCAN_WITH_IDENTITY) &&
+			    atomic_test_bit(bt_dev.flags, BT_DEV_SCANNING))) {
+				err = bt_id_set_adv_private_addr(adv);
+			}
```
I'm not sure I see what preemption issue I'm introducing that isn't already there, but anyway, what are you proposing?

olivier-le-sage committed:
bluetooth: host: Handle failure to disable scan when updating own_addr

It wasn't taken into account that bt_le_scan_set_enable()
has a return value. It's not likely that the controller
rejects the command when BT_DEV_SCANNING is set, however.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>

olivier-le-sage committed:
bluetooth: host: Do not try to set NRPA when scanning with identity

Attempting this would fail (assuming the controller is implemented
correctly) because when using legacy commands it is not allowed to
change the device address while scanning.

It also did not make sense. If we have configured the scanner to use
the identity address as own_addr, because the advertiser and scanner
addresses are shared when using legacy commands, setting the adv NRPA
here would overwrite the identity address used by the scanner, which
I assume is not the intention.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>

Thalley reviewed (dismissed):
LGTM

olivier-le-sage commented on subsys/bluetooth/host/id.c:
```diff
@@ -1967,23 +1969,39 @@ int bt_id_set_adv_own_addr(struct bt_le_ext_adv *adv, uint32_t options,
+			if (scan_disabled) {
 				bt_le_scan_set_enable(BT_HCI_LE_SCAN_ENABLE);
 			}
+#else
+			err = bt_id_set_adv_private_addr(adv);
+			*own_addr_type = BT_HCI_OWN_ADDR_RANDOM;
```
this function is `bt_id_set_adv_own_addr()`, not `bt_id_set_scan_own_addr()`

olivier-le-sage commented on subsys/bluetooth/host/id.c:
```diff
@@ -1789,25 +1789,36 @@ int bt_id_set_create_conn_own_addr(bool use_filter, uint8_t *own_addr_type)
 
 #if defined(CONFIG_BT_OBSERVER)
-static bool is_adv_using_rand_addr(void)
+static struct bt_le_ext_adv *get_adv(void)
```
Do you have a suggestion?

Thalley replied:
Hmm, it's not easy to be honest :D 

The functions, `is_legacy_adv_enabled` and `is_legacy_adv_using_id_addr` are also somewhat misleading, because they not considering legacy advertising sets when extended advertising is enabled (they specifically only check if extended advertising is disabled in both host and controller). 

Maybe something like instead of 
```c
static bool is_legacy_adv_enabled(void)
{
	struct bt_le_ext_adv *adv = get_legacy_adv();

	return adv != NULL && atomic_test_bit(adv->flags, BT_ADV_ENABLED);
}
```

```c
#define EXT_ADV_SUPPORTED IS_ENABLED(CONFIG_BT_BROADCASTER) && IS_ENABLED(CONFIG_BT_EXT_ADV) && BT_DEV_FEAT_LE_EXT_ADV(bt_dev.le.features))

static bool is_legacy_adv_enabled(void)
{
	if (!EXT_ADV_SUPPORTED) {
		return false;
	}

	struct bt_le_ext_adv *adv = bt_le_adv_lookup_legacy();

	return adv != NULL && atomic_test_bit(adv->flags, BT_ADV_ENABLED);
}
```

to avoid the `get_legacy_adv` completely? Could also implement `EXT_ADV_SUPPORTED` as a function instead

Looking over the codebase, we actually perform the `EXT_ADV_SUPPORTED` check many places, so it may actually be nice to define it as a complete check anyways. Not sure if the linker can optimize properly if it is turned into a function instead of using `IS_ENABLED`.

Thalley replied:
> Ok. What I'm seeing is you prefer to duplicate the code over using a helper function. I'll push that

That's one way to think of it :D I'd like to think of it as explicit and readable code over implied checks ;)

olivier-le-sage committed:
bluetooth: host: Fix uninitialized own_addr_type for legacy scan+adv

In 25c993e5b74562137c2951db01c8a8d200e98d3d a new case was introduced
where own_addr_type is not set by bt_id_set_scan_own_addr properly.

This led to issues for users where increasing their zephyr version
led to failures to start scanning after advertising in the case where
CONFIG_BT_SCAN_WITH_IDENTITY=n and legacy advertising commands are used.

Signed-off-by: Olivier Lesage <olivier.lesage@nordicsemi.no>
